### DISCUSSION TITLE
What exactly is a "module"?

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/runtime#101571] [Q&A]
I've seen the term used in .Net since day one, but I've never really understood what it is exactly, unlike an `Assembly` which has various classes representing it, a `Module` seems more nebulous, I've no idea what it is or why the term is even used.

**Comment by CyrusNajmabadi** [comment]
https://learn.microsoft.com/en-us/dotnet/standard/assembly/

> In .NET and .NET Framework, you can build an assembly from one or more source code files. In .NET Framework, assemblies can contain one or more modules.

https://learn.microsoft.com/en-us/dotnet/framework/app-domains/build-multifile-assembly

**Reply by CyrusNajmabadi** [reply]
> unlike an Assembly which has various classes representing it

This is not really accurate.  An assembly always has at least one module, and *it* contains those classes.  It can have more modules stitched together.  You'll see this in reflection with things like the ModuleBuilder type.  It's also exposed in Roslyn in our apis as well.

> I've no idea what it is or why the term is even used.

Consider coming to our discords for these sorts of free form questions.  You'll get responses and links normally in minutes or seconds.

**Comment by huoyaoyuan** [comment]
Historically, there was an idea that an assembly can be distributed as multiple files via Internet, and only executed part will be downloaded. The idea was then abandoned in the new century because directly executing code from Internet is totally horrible.

**Comment by The-Futurist** [comment]
> > unlike an Assembly which has various classes representing it
> 
> This is not really accurate. An assembly always has at least one module, and _it_ contains those classes. It can have more modules stitched together. You'll see this in reflection with things like the ModuleBuilder type. It's also exposed in Roslyn in our apis as well.
> 

It is accurate actually, an assembly **is** _represented_ by the [Assembly](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assembly?view=net-8.0) class and numerous other Assembly related classes.

**Reply by CyrusNajmabadi** [reply]
As does a module: 

https://learn.microsoft.com/en-us/dotnet/api/system.reflection.module?view=net-8.0
https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getmodules?view=net-8.0
https://learn.microsoft.com/en-us/dotnet/api/system.type.module?view=net-8.0

Etc :)

You're thinking about the flattened view of things.  But i'm explaining that that flattened view is precisely flattened from teh module-based view that it is actually can be constructed from :)  

Since you asked what modules were, it seems like getting this information would be valuable since it explains just that :)
